src/App.tsx

```diff
@@ -14,11 +16,149 @@ const goodsFromServer: string[] = [
   'Garlic',
 ];
 
-const App: React.FC = () => (
-  <div className="App">
-    <h1>Goods</h1>
-    {goodsFromServer.length}
-  </div>
-);
+type State = {
+  goodsList: string[],
+  reverseList: boolean,
+  start: boolean,
+  sortBy: string,
+  selectValue: number,
+};
+
+class App extends React.Component<{}, State> {
+  state: State = {
+    goodsList: [],
+    start: false,
+    reverseList: false,
+    sortBy: '',
+    selectValue: 1,
+  };
+
+  start = () => {
+    this.setState({
+      goodsList: goodsFromServer,
+      start: true,
+    });
+  };
+
+  reverse = () => {
+    this.setState(state => ({
+      reverseList: !state.reverseList,
+    }));
+  };
+
+  sortAlphabet = () => {
+    this.setState(() => ({
+      sortBy: 'alphabet',
+    }));
+  };
+
+  sortLength = () => {
+    this.setState(() => ({
+      sortBy: 'length',
+    }));
+  };
+
+  reset = () => {
+    this.setState({
+      reverseList: false,
+      sortBy: '',
+      selectValue: 1,
+    });
+  };
+
+  select = (event: React.ChangeEvent<HTMLSelectElement>) => {
+    this.setState({
+      selectValue: +event.target.value,
+    });
+  };
+
+  render() {
+    const {
+      goodsList,
+      start,
+      reverseList,
+      sortBy,
+      selectValue,
+    } = this.state;
+
+    const visibleList: string[] = goodsList.filter(
+      good => good.length >= selectValue,
+    );
+
+    visibleList.sort((g1, g2) => {
+      switch (sortBy) {
+        case 'alphabet':
+          return g1.localeCompare(g2);
+        case 'length':
+          return g1.length - g2.length;
+        default:
+          return 0;
+      }
+    });
+
+    if (reverseList) {
+      visibleList.reverse();
+    }
+
+    return (
+      <div className="App">
+        <h1>Goods</h1>
+        <button
+          type="button"
+          className={start ? 'button__hidden' : 'button'}
+          onClick={this.start}
+        >
+          Start
+        </button>
+
+        <button
+          type="button"
+          className={!start ? 'button__hidden' : 'button'}
+          onClick={this.reverse}
+        >
+          Revers
+        </button>
+
+        <button
+          type="button"
+          className={!start ? 'button__hidden' : 'button'}
+          onClick={this.sortAlphabet}
+        >
+          Sort alphabet
+        </button>
+
+        <button
+          type="button"
+          className={!start ? 'button__hidden' : 'button'}
+          onClick={this.sortLength}
+        >
+          Sort length
+        </button>
+
+        <button
+          type="button"
+          className={!start ? 'button__hidden' : 'button'}
+          onClick={this.reset}
+        >
+          Reset
+        </button>
+
+        <button type="button" className={start ? 'button' : 'button__hidden'}>
+          <label>
+            {'Select '}
+            <select
+              value={selectValue}
+              onChange={this.select}
+            >
+              <SelectedList />
+            </select>
```

Comment:
better to add this code inside SelectedList component

Reply:
Whole code?
I do not understand why?
All actions we made in the apps (onClick on Change) and the counter selectValue is outside.